pilot/pkg/proxy/envoy/v2/grpc.go

```diff
@@ -0,0 +1,262 @@
+// and can't understand normal Istio RDS - in particular expects "" instead of "/" as
+// default prefix, and is expects just the route for one host.
+
+func (s *DiscoveryServer) handleLDSApiType(con *XdsConnection, req *xdsapi.DiscoveryRequest) error {
+	if s.handleAck(con, req) {
+		return nil
+	}
+
+	push := s.globalPushContext()
+	resp := &xdsapi.DiscoveryResponse{
+		TypeUrl:     ListenerType,
+		VersionInfo: version,
+		Nonce:       nonce(push.Version),
+	}
+	var err error
+	for _, name := range req.ResourceNames {
+		portI := 0
+		hostn := name
+		port := ""
+		if strings.Contains(name, ":") {
+			hostn, port, err = net.SplitHostPort(name)
+			if err != nil {
+				return err
+			}
+			if port != "" {
+				portI, err = strconv.Atoi(port)
+				if err != nil {
+					return err
+				}
+			}
+		}
+		ll := &xdsapi.Listener{
+			Name: name,
+		}
+		// TODO: optimize, use a map in pre-computation of sidecar scope
+		for _, el := range con.node.SidecarScope.EgressListeners {
+			for _, s := range el.Services() {
+				if s.Hostname.Matches(host.Name(hostn)) {
+					ll.Address = &envoycore.Address{
+						Address: &envoycore.Address_SocketAddress{
+							SocketAddress: &envoycore.SocketAddress{
+								Address: s.Address,
+								PortSpecifier: &envoycore.SocketAddress_PortValue{
+									PortValue: uint32(portI),
+								},
+							},
+						},
+					}
+					ll.ApiListener = &envoy_config_listener_v2.ApiListener{
+						ApiListener: util.MessageToAny(&v2.HttpConnectionManager{
+							RouteSpecifier: &v2.HttpConnectionManager_Rds{
+								Rds: &v2.Rds {
+									RouteConfigName: name,
+								},
+							},
+						}),
+					}
+				}
+			}
+		}
+
+		lr := util.MessageToAny(ll)
+		resp.Resources = append(resp.Resources, lr)
+	}
+
+	err = con.send(resp)
+	if err != nil {
+		adsLog.Warnf("LDS: Send failure %s: %v", con.ConID, err)
+		recordSendError(ldsSendErrPushes, err)
+		return err
+	}
+
+	return nil
+}
+
+// Handle a gRPC CDS request, used with the 'ApiListener' style of requests.
+// The main difference is that the request includes Resources.
+func (s *DiscoveryServer) handleAPICDS(con *XdsConnection, req *xdsapi.DiscoveryRequest) bool {
+	if s.handleAck(con, req) {
+		return true
+	}
+
+	push := s.globalPushContext()
+	resp := &xdsapi.DiscoveryResponse{
+		TypeUrl:     ClusterType,
+		VersionInfo: version,
+		Nonce:       nonce(push.Version),
+	}
+
+	// gRPC doesn't currently support any of the APIs - returning just the expected EDS result.
+	// Since the code is relatively strict - we'll add info as needed.
+	for _, n := range req.ResourceNames {
+		hn, portn, err := net.SplitHostPort(n)
+		if err != nil {
+			adsLog.Warna("Failed to parse ", n, " ", err)
+			continue
+		}
+		rc := &xdsapi.Cluster{
+			Name: n,
+			ClusterDiscoveryType: &xdsapi.Cluster_Type{Type: xdsapi.Cluster_EDS},
+			EdsClusterConfig: &xdsapi.Cluster_EdsClusterConfig{
+				ServiceName: "outbound|" + portn + "||" + hn,
```

Comment:
not yet, and it'll be very different - i.e. a 'real' Listener since it'll need port, certs, etc.

The 'outbound' is not a real listener - gRPC doesn't capture or filter incoming traffic, it originates.

Reply:
Also I don't think we generate CDS for inbound clusters even in regular Istio.